csrc/ops/arith.cpp

```diff
@@ -2684,4 +2684,57 @@ TensorView* tensor(Val* val) {
   return out;
 }
 
+namespace {
+
+//! Create new output for matmul
+static TensorView* newForMatmul(
+    TensorView* tv_a,
+    TensorView* tv_b
+) {
+  auto orig_domain_a = TensorDomain::noReductions(tv_a->getMaybeRFactorDomain());
+  auto orig_domain_b = TensorDomain::noReductions(tv_b->getMaybeRFactorDomain());
+  auto ndims_a = orig_domain_a.size();
+  auto ndims_b = orig_domain_b.size();
+
+  NVF_ERROR(ndims_a>=1 && ndims_b>= 1);
+
+  std::vector<IterDomain*> new_domain;
+  auto higher_dim_domain = ndims_a >= ndims_b ? orig_domain_a : orig_domain_b;  
+  new_domain.reserve(higher_dim_domain.size());
+
+  if (ndims_a > 2 || ndims_b > 2) {
+    for (auto inx: c10::irange(higher_dim_domain.size() - 2)) {
+      new_domain.push_back(IterDomainBuilder(higher_dim_domain[inx]).resetSchedulingParams().build());
+    }
+  }
+  
+  // Add M domain to output domain if present
+  if (orig_domain_a.size() > 1) {
+    const IterDomain* m_id = orig_domain_a[ndims_a-2];
+    new_domain.push_back(IterDomainBuilder(m_id).resetSchedulingParams().build());
+  }
+
+  // Add N domain to output domain if present
+  if (orig_domain_b.size() > 1) {
+    const IterDomain* n_id = orig_domain_b[ndims_b-1];
+    new_domain.push_back(IterDomainBuilder(n_id).resetSchedulingParams().build());
+  }
+
+  TensorDomain* td = IrBuilder::create<TensorDomain>(
+      new_domain, TensorDomain::getContiguityFilledWith(new_domain, true));
+
+  return IrBuilder::create<TensorView>(td, tv_a->dtype());
+}
+
+} // namespace
+
+TensorView* eagerMatmul(
```

Comment:
Q:  Named it `eagerMatmul` to avoid clash with existing `matmul`. What should we name these new APIs? Preferably a consistent naming scheme across IR node/ API / scheduler. For instance: prefix `ATen`/`eager`/`exprEval`.

Reply:
I actually think it should just be `matmul`. We might translate these to nvfuser kernels after all...

Reply:
What is the case for keeping the original `matmul` and `linear`?

Reply:
There is an existing `matmul` here: https://github.com/NVIDIA/Fuser/blob/bfc8a1f2e4c39ba8e89ed1d4ee78b2940fbe27b3/csrc/ops/composite.cpp#L60. We can remove/rename that.

Reply:
Yes and that `matmul` should create a `MatmulOp` node. The code that's in there now could be converted to a utility in the future for decomposing a `MatmulOp`; that could be used in the nvfuser matmul scheduler.

Reply:
I will rename once we are ready to swap the backend of the python API to use the new nodes. Now that matmul is enabled in Thunder, renaming will break things there.

Reply:
Sounds good. Re thunder: they use the python API anyway so changes to these function names don't necessarily impact them, as long as the python frontend `matmul` op is updated accordingly.